### PR TITLE
feat: support spark submit jar on k8s

### PR DESCRIPTION
**Pull request by Zhao-LX2000**
### What is the purpose of the change

Support submit spark jar job on k8s.

### Environment
Spark：v3.2.1
Kubernetes: v1.23
```
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"23", GitVersion:"v1.23.5", GitCommit:"c285e781331a3785a7f436042c65c5641ce8a9e9", GitTreeState:"clean", BuildDate:"2022-03-16T15:58:47Z", GoVersion:"go1.17.8", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"23", GitVersion:"v1.23.17", GitCommit:"953be8927218ec8067e1af2641e540238ffd7576", GitTreeState:"clean", BuildDate:"2023-02-22T13:27:46Z", GoVersion:"go1.19.6", Compiler:"gc", Platform:"linux/amd64"}
```
Linkis：
image

### Test Command
```
linkis-cli --mode once \
-code 'sparkjars' \
-engineType spark-3.2.1 \
-codeType jar \
-labelMap engineConnMode=once \
-sourceMap jobName=OnceJobTest \
-k8sCluster 'K8S-default' \
-jobContentMap spark.app.main.class='org.apache.spark.examples.SparkPi' \
-jobContentMap runType='jar' \
-confMap spark.master='k8s-native' \
-confMap linkis.spark.k8s.serviceAccount='spark' \
-confMap spark.submit.deployMode='cluster' \
-confMap linkis.spark.k8s.master.url='k8s://https://172.31.226.155:6443' \
-confMap linkis.spark.k8s.config.file='/home/hadoop/.kube/config' \
-confMap linkis.spark.k8s.imagePullPolicy='IfNotPresent' \
-confMap spark.app.name='spark-submit-jar-k8s-test-linkis' \
-confMap spark.app.resource='local:///opt/spark/examples/jars/spark-examples_2.12-3.2.1.jar' \
-confMap linkis.spark.k8s.namespace='default'
```

### Brief change log

- SparkConfig.java: add `spark.kubernetes.driver.request.cores` and `spark.kubernetes.executor.request.cores` related fields.
- ClusterDescriptorAdapterFactory.java: add new k8s submission way("`k8s-native`", different from "`k8s-operator`").
- KubernetesApplicationClusterDescriptorAdapter.java: main features are as below
    - SparkLauncher add k8s related params setting.
    - Manually specifying driver pod name for follow-up task monitoring.
    - `initJobId()` method converts the state of the driver pod to the `SparkAppHandle.State` to determine whether the task has started.
- SparkOnKubernetesSubmitOnceExecutor.scala: new once executor type for k8s.
- SparkEngineConnFactory.scala: add `linkis.spark.k8s.driver.request.cores` and `linkis.spark.k8s.executor.request.cores` to sparkConf.
- SparkEngineConnResourceFactory.scala: According to [Spark Doc](https://spark.apache.org/docs/latest/running-on-kubernetes.html#configuration) , `spark.kubernetes.executor.request.cores` and `spark.kubernetes.driver.request.cores` have priority over `spark.executor.cores` and `spark.driver.cores`. So if `spark.kubernetes.executor.request.cores` is not specified when check resources, manually assign the value of `spark.executor.cores` to `spark.kubernetes.executor.request.cores` to make sure the default value of `spark.kubernetes.executor.request.cores` won't affect the tasks.
ca39e93103aa05ce81391ef3f8566c6

- SparkOnceExecutorFactory.scala: Create `SparkOnKubernetesSubmitOnceExecutor` if `-k8sCluster` is specified.
- SparkJobProgressUtil.scala: main features are as below
    - Methods add `podIP` params to be compatible with the spark tasks on k8s.
    - `getKubernetesSparkJobInfo()`: method for getting status of spark tasks through Spark UI exposed on driver pod’s 4040 port.

### Checklist

- [x] I have read the [Contributing Guidelines on pull requests](https://github.com/facebook/docusaurus/blob/main/CONTRIBUTING.md#pull-requests).
- [x] I have explained the need for this PR and the problem it solves
- [x] I have explained the changes or the new features added to this PR
- [ ] I have added tests corresponding to this change
- [ ] I have updated the documentation to reflect this change
- [ ] I have verified that this change is backward compatible (If not, please discuss on the [Linkis mailing list](https://linkis.apache.org/community/how-to-subscribe) first)
- [ ] **If this is a code change**: I have written unit tests to fully verify the new behavior.

### CODE REVIEW

File: linkis-engineconn-plugins/spark/src/main/scala/org/apache/linkis/engineplugin/spark/executor/SparkOnKubernetesSubmitOnceExecutor.scala [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+  }
+
+  override def getProgress: Float = {
+    if (clusterDescriptorAdapter == null) {
+      logger.info("clusterDescriptorAdapter is null")
+    } else if (clusterDescriptorAdapter.getJobState == null) {
+      logger.info("clusterDescriptorAdapter.getJobState is null")
+    } else {
+      logger.info("clusterDescriptorAdapter/getJobState is not null")
+    }
```
Comment (chenghuichen) [inline linkis-engineconn-plugins/spark/src/main/scala/org/apache/linkis/engineplugin/spark/executor/SparkOnKubernetesSubmitOnceExecutor.scala:135]: These logs seem to be used for debugging. May be too noisy in prod env.

File: linkis-engineconn-plugins/spark/src/main/java/org/apache/linkis/engineplugin/spark/client/deployment/KubernetesApplicationClusterDescriptorAdapter.java [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+    SparkConfig sparkConfig = executionContext.getSparkConfig();
+    sparkLauncher = new CustomSparkSubmitLauncher();
+    sparkLauncher
+        .setJavaHome(sparkConfig.getJavaHome())
+        .setSparkHome(sparkConfig.getSparkHome())
+        .setMaster(sparkConfig.getK8sMasterUrl())
+        .setDeployMode(sparkConfig.getDeployMode())
+        .setAppName(sparkConfig.getAppName())
+        .setVerbose(true);
+    this.driverPodName = sparkConfig.getAppName() + "-" + System.currentTimeMillis() + "-driver";
```
Comment (chenghuichen) [inline linkis-engineconn-plugins/spark/src/main/java/org/apache/linkis/engineplugin/spark/client/deployment/KubernetesApplicationClusterDescriptorAdapter.java:71]: Using timestamp to make driverPodName unique is not enough good, UUID would be better. 
If this is for pod create time recording, there is creationTimestamp in pod spec already.

File: linkis-engineconn-plugins/spark/src/main/java/org/apache/linkis/engineplugin/spark/client/deployment/ClusterDescriptorAdapterFactory.java [hunk 3]
```diff
@@ -29,8 +29,13 @@ public static ClusterDescriptorAdapter create(ExecutionContext executionContext)
     ClusterDescriptorAdapter clusterDescriptorAdapter =
         new YarnApplicationClusterDescriptorAdapter(executionContext);
 
-    if (StringUtils.isNotBlank(master) && master.equalsIgnoreCase("k8s-operator")) {
-      clusterDescriptorAdapter = new KubernetesOperatorClusterDescriptorAdapter(executionContext);
+    if (StringUtils.isNotBlank(master)) {
+      if (master.equalsIgnoreCase("k8s-operator")) {
+        clusterDescriptorAdapter = new KubernetesOperatorClusterDescriptorAdapter(executionContext);
+      } else if (master.equalsIgnoreCase("k8s-jar")) {
```
Comment (chenghuichen) [inline linkis-engineconn-plugins/spark/src/main/java/org/apache/linkis/engineplugin/spark/client/deployment/ClusterDescriptorAdapterFactory.java:35]: In my opinion, "k8s-native" would be more suitable than "k8s-jar", which is aligned to the concept "k8s-operator".

File: linkis-engineconn-plugins/spark/src/main/scala/org/apache/linkis/engineplugin/spark/executor/SparkOnKubernetesSubmitOnceExecutor.scala (hunk 1 above)
Comment (chenghuichen) [inline linkis-engineconn-plugins/spark/src/main/scala/org/apache/linkis/engineplugin/spark/executor/SparkOnKubernetesSubmitOnceExecutor.scala:135]: These logs seem to be used for debugging. May be too noisy in prod env.

File: linkis-engineconn-plugins/spark/src/main/scala/org/apache/linkis/engineplugin/spark/factory/SparkOnceExecutorFactory.scala [hunk 4]
```diff
@@ -34,11 +40,18 @@ class SparkOnceExecutorFactory extends OnceExecutorFactory {
       engineCreationContext: EngineCreationContext,
       engineConn: EngineConn,
       labels: Array[Label[_]]
-  ): OnceExecutor =
+  ): OnceExecutor = {
+    val clusterLabel = LabelUtil.getLabelFromArray[ClusterLabel](labels)
     engineConn.getEngineConnSession match {
+      case context: SparkEngineConnContext
```
Comment (peacewong) [inline linkis-engineconn-plugins/spark/src/main/scala/org/apache/linkis/engineplugin/spark/factory/SparkOnceExecutorFactory.scala:46]: It is recommended to optimize it into an if else judgment

Comment (Zhao-LX2000) [reply linkis-engineconn-plugins/spark/src/main/scala/org/apache/linkis/engineplugin/spark/factory/SparkOnceExecutorFactory.scala]: Already done in the last commitment.

**Review by guoshupei — COMMENTED** [review]
LGTM

**Review by peacewong — APPROVED** [review]
LGTM.

**Pull request merged**